jux/map/board.py

```diff
@@ -88,18 +88,27 @@ def ore(self) -> Array:
     @property
     def valid_spawns_mask(self) -> Array:  # bool[height, width]
         valid_spawns_mask = (~self.map.ice & ~self.map.ore)  # bool[height, width]
-        valid_spawns_mask = valid_spawns_mask & jnp.roll(valid_spawns_mask, 1, axis=0)
-        valid_spawns_mask = valid_spawns_mask & jnp.roll(valid_spawns_mask, -1, axis=0)
+        batched = True
```

Comment:
Same as below. There may be more than one batch dimension, but batch dimensions are always leading dimension, so we can always access the dimensions we want by negative index. Here is an example of how to tackle more general cases without special treatment for some cases.
```python
    @property
    def valid_spawns_mask(self) -> Array:  # bool[height, width]
        valid_spawns_mask = (~self.map.ice & ~self.map.ore)  # bool[..., height, width]
        valid_spawns_mask = valid_spawns_mask & jnp.roll(valid_spawns_mask, 1, axis=-2)
        valid_spawns_mask = valid_spawns_mask & jnp.roll(valid_spawns_mask, -1, axis=-2)
        valid_spawns_mask = valid_spawns_mask & jnp.roll(valid_spawns_mask, 1, axis=-1)
        valid_spawns_mask = valid_spawns_mask & jnp.roll(valid_spawns_mask, -1, axis=-1)
        valid_spawns_mask = valid_spawns_mask.at[..., [0, -1], :].set(False)
        valid_spawns_mask = valid_spawns_mask.at[..., :, [0, -1]].set(False)

        batch_shape = factory_overlap.shape[:-3]
        factory_overlap = self.factory_pos[..., None, :] + delta_xy  # int[..., 2 * MAX_N_FACTORIES, 85, 2]
        factory_overlap = factory_overlap.reshape(batch_shape + (-1, 2))
        factory_overlap = jnp.clip(factory_overlap, 0, jnp.array([self.height - 1, self.width - 1]))
        valid_spawns_mask = valid_spawns_mask.at[factory_overlap[..., 0], factory_overlap[..., 1]].set(False, mode='drop')
```

Reply:
this still has a bug with factory_overlap[...] and setting those points to False on valid_spawns_mask.

In batch, valid_spawns_mask is shape `(b, 48, 48)`
factory_overlap is shape `(b, 1020, 2)`

`valid_spawns_mask.at[factory_overlap[..., 0], factory_overlap[..., 1]]` will incorrectly select the wrong indices. `factory_overlap[..., 0]` will select the batch index, and `factory_overlap[..., 1]` will select the first map dimension.